File: apps/mobile/ChinaTextbookStudy/Domain/Quests.swift
import Foundation

/// Daily quests — three per day, chosen deterministically from the date so the
/// set is stable for a given day without any server involvement.

enum QuestKind: String, Codable, Hashable, CaseIterable {
    case earnXP        // earn N XP today
    case finishLessons // complete N lessons today
    case reviewMistakes// review N mistakes today
    case readTexts     // finish N passages/stories today

    var symbol: String {
        switch self {
        case .earnXP:         return "bolt.fill"
        case .finishLessons:  return "checkmark.seal.fill"
        case .reviewMistakes: return "arrow.triangle.2.circlepath"
        case .readTexts:      return "book.fill"
        }
    }
}

struct Quest: Identifiable, Hashable {
    let kind: QuestKind
    let target: Int
    let reward: Int          // gems

    /// Stable within a day — the claim ledger keys off `"\(date):\(id)"`.
    var id: String { "\(kind.rawValue)-\(target)" }

    var title: String {
        switch kind {
        case .earnXP:         return "获得 \(target) 点经验"
        case .finishLessons:  return "完成 \(target) 节小课"
        case .reviewMistakes: return "复习 \(target) 道错题"
        case .readTexts:      return "读完 \(target) 篇课文或故事"
        }
    }
}

enum Quests {
    /// Candidate pool. Kept small and achievable — these are daily habits,
    /// not grinds.
    private static let pool: [Quest] = [
        Quest(kind: .earnXP, target: 30, reward: 10),
        Quest(kind: .earnXP, target: 60, reward: 15),
        Quest(kind: .earnXP, target: 100, reward: 25),
        Quest(kind: .finishLessons, target: 1, reward: 10),
        Quest(kind: .finishLessons, target: 2, reward: 20),
        Quest(kind: .finishLessons, target: 3, reward: 30),
        Quest(kind: .reviewMistakes, target: 1, reward: 10),
        Quest(kind: .reviewMistakes, target: 3, reward: 20),
        Quest(kind: .readTexts, target: 1, reward: 15),
        Quest(kind: .readTexts, target: 2, reward: 25),
    ]

    /// SplitMix64 finalizer. A plain rolling hash leaves the high bits nearly
    /// identical for consecutive dates (the last character only feeds the low
    /// bits), which collapsed the variety to a handful of sets per month — so
    /// every derived index gets avalanched through this first.
    private static func mix(_ value: UInt64) -> UInt64 {
        var x = value &+ 0x9E37_79B9_7F4A_7C15
        x = (x ^ (x >> 30)) &* 0xBF58_476D_1CE4_E5B9
        x = (x ^ (x >> 27)) &* 0x94D0_49BB_1331_11EB
        return x ^ (x >> 31)
    }

    /// Three quests of distinct kinds for the given day, derived from the date
    /// string so every launch on the same day yields the same set.
    static func daily(for dateString: String) -> [Quest] {
        var hash = UInt64(5381)
        for byte in dateString.utf8 { hash = hash &* 33 &+ UInt64(byte) }
        hash = mix(hash)

        func index(_ count: Int, salt: UInt64) -> Int {
            guard count > 0 else { return 0 }
            return Int(mix(hash &+ salt) % UInt64(count))
        }
        func pick(_ kind: QuestKind, salt: UInt64) -> Quest {
            let options = pool.filter { $0.kind == kind }
            return options[index(options.count, salt: salt)]
        }

        // Always an XP quest, then two other kinds — drawn without replacement
        // so a day never offers the same kind twice.
        var rest: [QuestKind] = [.finishLessons, .reviewMistakes, .readTexts]
        let firstKind = rest.remove(at: index(rest.count, salt: 0x11))
        let secondKind = rest[index(rest.count, salt: 0x22)]

        return [
            pick(.earnXP, salt: 0x01),
            pick(firstKind, salt: 0x33),
            pick(secondKind, salt: 0x44),
        ]
    }
}

File: packages/core/src/quests.ts
/**
 * quests.ts —— 每日任务（Daily Quests）单一事实源
 *
 * 与 iOS `Domain/Quests.swift` 逐行对齐：同一天双端必须给出完全相同的
 * 三条任务（同 kind / 同 target / 同 reward / 同顺序）。
 *
 * 确定性来源：日期字符串（YYYY-MM-DD）→ djb2 滚动哈希 → SplitMix64
 * finalizer 雪崩 → 对候选池取模。哈希原语抽在 rng.ts（quests / league 共用），
 * TS 端用 BigInt 模拟 Swift 的 UInt64 环绕运算（&+ / &*），位运算逐位一致；
 * spec/golden-vectors.json 的 `quests` 组是双端对照的黄金向量。
 */

import { U64_MASK, djb2Hash, mix64 as mix } from "./rng";

export type QuestKind = "earnXP" | "finishLessons" | "reviewMistakes" | "readTexts";

export interface Quest {
  kind: QuestKind;
  /** 目标计数（earnXP 为 XP 数，其余为次数） */
  target: number;
  /** 完成奖励（宝石） */
  reward: number;
  /** 当日内稳定的 id —— 领取账本以 `"${date}:${id}"` 记账 */
  id: string;
  /** 儿童友好中文标题（与 iOS 文案一致） */
  title: string;
}

function makeQuest(kind: QuestKind, target: number, reward: number): Quest {
  return { kind, target, reward, id: `${kind}-${target}`, title: questTitle(kind, target) };
}

/** 与 iOS `Quest.title` 完全一致的中文文案。 */
export function questTitle(kind: QuestKind, target: number): string {
  switch (kind) {
    case "earnXP":         return `获得 ${target} 点经验`;
    case "finishLessons":  return `完成 ${target} 节小课`;
    case "reviewMistakes": return `复习 ${target} 道错题`;
    case "readTexts":      return `读完 ${target} 篇课文或故事`;
  }
}

/**
 * 候选池 —— 与 iOS `Quests.pool` 逐项一致（顺序也一致，
 * 因为 pick 是「按 kind 过滤后取模」，顺序参与结果）。
 */
export const QUEST_POOL: readonly Quest[] = [
  makeQuest("earnXP", 30, 10),
  makeQuest("earnXP", 60, 15),
  makeQuest("earnXP", 100, 25),
  makeQuest("finishLessons", 1, 10),
  makeQuest("finishLessons", 2, 20),
  makeQuest("finishLessons", 3, 30),
  makeQuest("reviewMistakes", 1, 10),
  makeQuest("reviewMistakes", 3, 20),
  makeQuest("readTexts", 1, 15),
  makeQuest("readTexts", 2, 25),
];

/**
 * 给定日期（YYYY-MM-DD）的三条每日任务，确定性生成：
 *   - 第一条必为 earnXP；
 *   - 后两条从 finishLessons / reviewMistakes / readTexts 中不放回抽取，
 *     因此一天内三条任务 kind 两两不同；
 *   - 同一天任何时刻、任何端调用结果完全一致。
 */
export function dailyQuests(dateString: string): Quest[] {
  const hash = mix(djb2Hash(dateString));

  const index = (count: number, salt: bigint): number => {
    if (count <= 0) return 0;
    return Number(mix((hash + salt) & U64_MASK) % BigInt(count));
  };
  const pick = (kind: QuestKind, salt: bigint): Quest => {
    const options = QUEST_POOL.filter(q => q.kind === kind);
    return options[index(options.length, salt)];
  };

  // 与 iOS 一致：先抽走 firstKind（不放回），再在剩余里抽 secondKind
  const rest: QuestKind[] = ["finishLessons", "reviewMistakes", "readTexts"];
  const [firstKind] = rest.splice(index(rest.length, 0x11n), 1);
  const secondKind = rest[index(rest.length, 0x22n)];

  return [pick("earnXP", 0x01n), pick(firstKind, 0x33n), pick(secondKind, 0x44n)];
}
